### ISSUE TITLE
restic repo initialised twice when starting up docker with both backup/pruneprune actually runs

### ISSUE THREAD

**Issue by sumnerboy12** [djmaze/resticker#48]
I noticed when I started my docker swarm stack which included both `backup` and `prune`, that it attempted to (and succeeded) in initialising the same repo twice. 

This put the repo in a bad state and any attempts to run a backup/prune resulted in;
```
Fatal: config cannot be loaded: ciphertext verification failed
```

Can we either only attempt to initialise during `backup`? In `prune` can we check if the repo is initialised and if not then exit early (clearly nothing to do).

**Comment by ThomDietrich** [comment]
How is this even possible?

Initialization happens outside of the backup or prune code inside `entrypoint` - could this be related to another aspect of your setup?

Also, whatever led to the issue might be resolved after PR #47 is merged. Would you be able to test the new logic with your setup?

**Comment by sumnerboy12** [comment]
I have a docker stack with two services, `backup` and `prune` (i.e. one with `BACKUP_CRON` and one with `PRUNE_CRON`). When I deployed the stack they both started up simultaneously on the same node against an uninitialised repo (on Backblaze). 

They both managed to initialise the repo at the same time and put it into an invalid state. This has been reported [here](https://github.com/restic/restic/issues/2323#issuecomment-509970420).

(BTW - I think I recognise your handle also - openhab perhaps?!)

**Comment by ThomDietrich** [comment]
Hey!

Which part of the issue behind your link talks about initialization? The main thing I get from the issue is that "this kind of issue" is normally linked to the storage and less to restic.

Let's clarify first that this is indeed a real issue, if so we need to discuss whether initialization should only happen with `BACKUP_CRON`.

Right. openHAB, of course! :)

**Comment by sumnerboy12** [comment]
Just the comment that talked about that error being caused by the repo being initialised twice (somehow). When I got this error I was searching for relevant posts and came across that one. I noticed when checking the docker logs for my `djmaze/resticker` services that both the `backup` and `prune` services had started up simultaneously and had logging to say they had both initialised the repo.

The next time either of them attempted to do anything, or I tried to access the repo manually (via `restic snapshots`), it got the `ciphertext validation` error.

I was pretty much following the example in https://github.com/djmaze/resticker/blob/master/docker-swarm.example.yml.

**Comment by zoispag** [comment]
> Hey!
> 
> 
> 
> Which part of the issue behind your link talks about initialization? The main thing I get from the issue is that "this kind of issue" is normally linked to the storage and less to restic. Can you verify this happens with the new logic and with simultaneous startup.
> 
> Regarding simultaneous: Why did you set it up like this? Wouldn't it then be better to add `--prune` to the forget command?
> 
> 
> 
> Let's clarify first that this is indeed a real issue, if so we need to discuss whether initialization should only happen with `BACKUP_CRON`.
> 
> 
> 
> Right. openHAB, of course! :)

@ThomDietrich the `—prune` is not realistic for big repos with multiple hosts running backup, since prune is a blocking action. That’s why @djmaze created the second container for pruning in #36

2 different containers (backup & prune) starting at the same time would indeed cause the initialization twice. When I tried it, all my repos were already initialized.

**Comment by ThomDietrich** [comment]
@zoispag that makes sense. Thanks.
Got it. I guess we can safely assume that a repo will already be initialized in the prune usecase of the image. Shall we separate the check in #47 then. Check whether repository reachable and initialized but only initialize in the prune case? @djmaze 

Also: in the example yml files the container start with a 30min difference. For the purpose of an example I would suggest to make it 12 hours.

**Comment by zoispag** [comment]
> @zoispag that makes sense. Thanks.
> 
> Got it. I guess we can safely assume that a repo will already be initialized in the prune usecase of the image. Shall we separate the check in #47 then. Check whether repository reachable and initialized but only initialize in the backup case? @djmaze 
> 
> 
> 
> Also: in the example yml files the container start with a 30min difference. For the purpose of an example I would suggest to make it 12 hours.

Pruning will start with 30min difference. The containers will start at the same time, if they are at the same docker-compose file.

**Comment by ThomDietrich** [comment]
Of course. Obviously I was talking about the cron job definition :)

**Comment by djmaze** [comment]
> Check whether repository reachable and initialized but only initialize in the backup case? @djmaze

That sounds reasonable. Let's do it.

**Comment by djmaze** [comment]
I just realized that, at least in a swarm deployment, multiple initialization can also occur when the `backup` service is deployed to multiple machines. So just skipping the initialization in the prune service does not completely alleviate this problem.

In my Docker swarm deployments, I sometimes make use of separate "one-shot" initialization services. They are just started once and then exit. On swarm, this is possible using a restart policy with `condition: on-failure`. When using Docker compose, it should be possible to use `restart: on-failure` instead. 

So we could have a separate `init` service which is just used for initializing the repository on first use.  Maybe this would be the way to go. What do you think?

**Comment by djmaze** [comment]
Of course, when using Docker compose, it would be enough to add an `init` *command* for the entrypoint so you would just need to call something like `docker-compose run backup init` once.

**Comment by ThomDietrich** [comment]
Before we discuss this any further: Is there any reason why restic should even allow this? Instead of building a workaround here maybe we can help come up with an improvement in restic itself!?

**Comment by ThomDietrich** [comment]
> Also: in the example yml files the container start with a 30min difference. For the purpose of an example I would suggest to make it 12 hours.

Moved to #49

**Comment by djmaze** [comment]
> Before we discuss this any further: Is there any reason why restic should even allow this? Instead of building a workaround here maybe we can help come up with an improvement in restic itself!?

Would be okay for me. Then we should at least give a hint about this in the documentation (as you thankfully already began in #49).

**Comment by sumnerboy12** [comment]
What about removing the auto-initialisation and requiring the user to do this step manually (with suitable documentation/instructions)?

**Comment by zoispag** [comment]
> What about removing the auto-initialisation and requiring the user to do this step manually (with suitable documentation/instructions)?

I don’t think that’s a good idea. I do rely on restiker initializing the repos for me.

On the other hand, we can add a mild delay on the prune container on startup, to wait a minute before attempting initialization.

**Comment by djmaze** [comment]
I don't really like both ideas. Initializing the repo manually yourself means you have to also setup all the SSH keys, credentials etc. locally, which is prone to errors. The 1 minute wait, on the other hand, would rather be an ugly workaround and does not protect when deploying multiple backup containers simultaneously.

Ideally, this should be solved on the restic side. Or for a more short-term solution, we could have a separate `init` command as proposed before.

**Comment by ThomDietrich** [comment]
I'm not strictly against the carved out `init` command (which is process wise what @sumnerboy12 said). In my opinion it is anyhow better to leave this one-time command to an intentional action by the user - even though I can't currently think of a scenario in which it might lead to unexpected behavior or security risk.

To summarize the solution:
1. In backup and prune "mode" the entrypoint script would check the connection to the repository and exit with an error if it is not accessible or initialized.
1. Initialization has to be executed manually with e.g. `docker-compose run backup init`
1. An optional environment variable (default off) allows the user to re-enable current auto-initialize behavior

In other words: The solution is a PR that is 10% code and 90% docu changes.

What do you think? Do we want to offer point 3?

**Comment by zoispag** [comment]
If init is removed from both backup/prune containers, i would definitely need an ENV variable, to turn auto-init on. I am using this as part of my automation.

On the other hand, does it make sense to init a repo to do prune? It should already exist. We can remove init from prune altogether. (This still does not solve the issue of multiple backup containers on swarm)

**Comment by camo-f** [comment]
Hello,

`init` is still an issue, I got the previously reported error with the current docker-compose example.

I think what @ThomDietrich summed up here is a good approach.
> To summarize the solution:
> 
>   1. In backup and prune "mode" the entrypoint script would check the connection to the repository and exit with an error if it is not accessible or initialized.
> 
>   2. Initialization has to be executed manually with e.g. `docker-compose run backup init`
> 
>   3. An optional environment variable (default off) allows the user to re-enable current auto-initialize behavior

I also agree with @zoispag that init shouldn't be done when pruning, as repo should already be initialised. So the optional env variable would only have effect in backup containers.

For Docker Swarm, @djmaze mentioned an initialisation service, which seems the cleanest way to handle multiple backup services to me. It could also be used for Docker Compose to only have a single file, which can be better for a quickstart.

@ThomDietrich Have you already worked on this issue ? It doesn't seem too complex to me, so I can help if needed.

**Comment by ThomDietrich** [comment]
Triggered by a linked project I will most probably take some time to work on this issue this week. I will respond to your thoughts and suggestions later. Thanks for the push

**Comment by varac** [comment]
Yes, please reopen this bug - I hit it every time I want to init a repo for a new host where both backup and prune containers are initializing at the same time.

**Comment by djmaze** [comment]
Oh, well, seems I overlooked the previous comment. Sorry

**Comment by varac** [comment]
Any news on this ? Right now, every new host added to our backup needs manual intervention, would be great to have this fixed.

**Comment by djmaze** [comment]
Sorry for the late answer. If there is no one on this, I will try to find some time to implement this in the upcoming days.

**Comment by MRezaNasirloo** [comment]
I faced this issue while trying to set up restic on a new host with an existing repo, no my repo is broken and not usable anymore
```
Fatal: config or key 6a88fbbb5b0fa776b220d4cff42dd5716a3f691b550ffbec5a5fb15f15b2153c is damaged: ciphertext verification failed. Try again
```

**Comment by pquerner** [comment]
I just had the same and temporary added a `profiles` entry to the service `prune` and `check` like described [here](https://stackoverflow.com/a/65922688).

After starting the service, only the `backup` service ran - and found no repository, therefore issued the `init` command on `restic`.
Then I stopped the containers - and removed the `profiles` config from the .yml file and started it again - this time it did not create multiple `keys` (etc).

Maybe this needs a `init` service and all services need to wait for this service?
